File: workers/research/src/black_book_research/adapters/federal/loc/definition.py
"""Library of Congress adapter definition (BB-046)."""

from __future__ import annotations

from black_book_research.adapters.types import RateLimitPolicy

from ..shared.contract_builder import (
    DEFAULT_FEDERAL_EXPORT_FILTER,
    DEFAULT_FEDERAL_RETENTION,
    FEDERAL_PUBLIC_DOMAIN_RIGHTS,
    build_federal_adapter_definition,
)
from ..shared.types import FederalRetentionRules

LOC_ADAPTER_ID = "loc-collections-v1"
LOC_PARSER_VERSION = "parser-1.0.0"

loc_adapter_definition = build_federal_adapter_definition(
    family="loc",
    adapter_id=LOC_ADAPTER_ID,
    parser_version=LOC_PARSER_VERSION,
    display_name="Library of Congress Collections",
    classification="primary_archival",
    stable_id_scheme="loc-lccn",
    source_id="src_loc",
    organization_id="org_loc",
    rights=FEDERAL_PUBLIC_DOMAIN_RIGHTS,
    permitted_claim_classes=("biographical_fact", "geographic_fact", "institutional_fact"),
    rate_limits=RateLimitPolicy(requests_per_minute=20, burst=3),
    expected_records_per_run=50,
    refresh_schedule="0 4 * * 2",
    retention=FederalRetentionRules(
        required_fields=DEFAULT_FEDERAL_RETENTION.required_fields,
        min_title_length=DEFAULT_FEDERAL_RETENTION.min_title_length,
        allowed_classifications=("primary_archival", "government_record"),
        require_canonical_url=DEFAULT_FEDERAL_RETENTION.require_canonical_url,
    ),
    export_filter=type(DEFAULT_FEDERAL_EXPORT_FILTER)(
        max_payload_bytes=DEFAULT_FEDERAL_EXPORT_FILTER.max_payload_bytes,
        strip_keys=DEFAULT_FEDERAL_EXPORT_FILTER.strip_keys + ("imageTiles", "marcRecord"),
        essential_keys=(
            "stableIdentifier",
            "title",
            "canonicalUrl",
            "classification",
            "lccn",
            "date",
            "subjects",
        ),
    ),
)

File: packages/domain/src/adapters/federal/loc/definition.ts
/**
 * Library of Congress adapter definition, rights, rate limits, and kill switch (BB-046).
 */
import {
  buildFederalAdapterDefinition,
  DEFAULT_FEDERAL_EXPORT_FILTER,
  DEFAULT_FEDERAL_RETENTION,
  FEDERAL_PUBLIC_DOMAIN_RIGHTS,
} from '../shared/contract-builder.js';

export const LOC_ADAPTER_ID = 'loc-collections-v1' as const;
export const LOC_PARSER_VERSION = 'parser-1.0.0' as const;

export const locAdapterDefinition = buildFederalAdapterDefinition({
  family: 'loc',
  adapterId: LOC_ADAPTER_ID,
  parserVersion: LOC_PARSER_VERSION,
  displayName: 'Library of Congress Collections',
  classification: 'primary_archival',
  stableIdScheme: 'loc-lccn',
  sourceId: 'src_loc',
  organizationId: 'org_loc',
  rights: FEDERAL_PUBLIC_DOMAIN_RIGHTS,
  permittedClaimClasses: ['biographical_fact', 'geographic_fact', 'institutional_fact'],
  rateLimits: { requestsPerMinute: 20, burst: 3 },
  expectedRecordsPerRun: 50,
  refreshSchedule: '0 4 * * 2',
  retention: {
    ...DEFAULT_FEDERAL_RETENTION,
    allowedClassifications: ['primary_archival', 'government_record'],
  },
  exportFilter: {
    ...DEFAULT_FEDERAL_EXPORT_FILTER,
    stripKeys: [...DEFAULT_FEDERAL_EXPORT_FILTER.stripKeys, 'imageTiles', 'marcRecord'],
    essentialKeys: ['stableIdentifier', 'title', 'canonicalUrl', 'classification', 'lccn', 'date', 'subjects'],
  },
});
